contracts/PaymentReceiver.sol

```diff
@@ -28,76 +28,114 @@ contract PaymentReceiver is IPaymentReceiver, Database {
     function deleteGate(uint256 _gateId) public override {
         Gate memory gate = getGate(_gateId);
 
-        for (uint256 index = 0; index < gate.activeUsers.length; index++) {
-            address addr = gate.activeUsers[index];
-            _deleteFlow(addr, gate.payee, onlyToken);
+        uint256 index = 0;
+        for(index; index < gate.activeUsers; index++) {
+            address addr = gateUsers[_gateId][index];
+            _deleteFlow(addr, gate.payee, gate.flowRate, onlyToken);
+            checkedIn[addr][_gateId] = false;
             emit CheckOut(addr , _gateId);
-        } 
+        }
+        gate.activeUsers = 0;
 
         super.deleteGate(_gateId);
     }
 
     function checkIn(uint256 _gateId) external {
-        Gate storage gate = gates[_gateId];
+        require(checkedIn[msg.sender][_gateId] == false, "you are already checked in");
 
-        for (uint256 index = 0; index < gate.activeUsers.length; index++) {
-            require(gate.activeUsers[index] != msg.sender, string(abi.encodePacked("already checked in at: ", gate.name)));
-        }
+        Gate storage gate = gates[_gateId];
 
         _createFlow(gate.payee, gate.flowRate, onlyToken);
-        gate.activeUsers.push(msg.sender);
+        checkedIn[msg.sender][_gateId] = true;
+        gateUsers[_gateId][gate.activeUsers] = msg.sender;
+        gate.activeUsers++;
```

Comment:
I think this line causes an error, since we are checking out the most recently checked in user here, where an older user might be the one actually checked out. I have added a test which fails due to this error.